src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,98 @@
+import cn from 'classnames';
+import React from 'react';
+
+import { deleteTodo } from '../../api/todos';
+import { useTodos } from '../../hooks/useTodos';
+import { Todo } from '../../types/Todo';
+
+type Props = {
+  todo: Todo;
+};
+
+const TodoItem: React.FC<Props> = ({ todo }) => {
+  const [selectedTodoId, setSelectedTodoId] = React.useState<number>(0);
+
+  const {
+    todos,
+    setTodos,
+    isLoading,
+    setIsLoading,
+    handleError,
+    isAllDeleted,
+  } = useTodos();
+
+  const handleCheckbox = () => {
+    setTodos(
+      todos.map(prevTodo => {
+        return prevTodo.id === todo.id
+          ? { ...prevTodo, completed: !prevTodo.completed }
+          : prevTodo;
+      }),
+    );
+  };
+
+  const handleDeleteTodo = () => {
+    setIsLoading(true);
+    setSelectedTodoId(todo.id);
+
+    deleteTodo(todo.id)
+      .then(() => {
+        setTodos(todos.filter(prevTodo => prevTodo.id !== todo.id));
+      })
+      .catch(() => {
+        handleError('Unable to delete a todo');
+      })
+      .finally(() => {
+        setIsLoading(false);
+      });
+  };
+
+  const firstCondition = isLoading && todo.id === selectedTodoId;
+  const secondCondition = todo.completed && isAllDeleted;
+  const isLoaderActive = firstCondition || secondCondition;
```

Comment:
That is strange and too complicated logic and these conditions will be more. You can create a state where you can store ids of todos that currently loading and depending on it set a loader.